projects/lesson-3/runtime/src/kitties.rs

```diff
@@ -1,37 +1,197 @@
+		        .ok_or("Overflow adding a new kitty to total supply")?;
+
+		    // `nonce` and `random_hash` generation can stay here
+		    let nonce = Nonce::get();
+		    let random_hash = (<system::Module<T>>::random_seed(), &sender, nonce)
+		        .using_encoded(<T as system::Trait>::Hashing::hash);	    
+			let payload = (<system::Module<T>>::random_seed(), &sender, 
+				<system::Module<T>>::extrinsic_index(), 
+				<system::Module<T>>::block_number());
+			let dna_hash = payload.using_encoded(blake2_128);
+
+
+		    // ACTION: Move this collision check to the `mint()` function
+		    ensure!(!<KittyOwner<T>>::exists(random_hash), "Kitty already exists");
+
+		    // Creating the `Kitty` object can stay here
+		    let new_kitty = Kitty {
+		        id: random_hash,
+		        dna: dna_hash,
+		        price: 0.into(),
+		        gen: 0,
+		    };
+
+
+		    // ACTION: Move all of the kitty related storage updates to the `mint()` function
+		    <Kitties<T>>::insert(random_hash, new_kitty);
+		    <KittyOwner<T>>::insert(random_hash, &sender);
+
+		    <AllKittiesArray<T>>::insert(all_kitties_count, random_hash);
+		    <AllKittiesCount>::put(new_all_kitties_count);
+		    <AllKittiesIndex<T>>::insert(random_hash, all_kitties_count);
+
+		    <OwnedKittiesArray<T>>::insert((sender.clone(), owned_kitty_count), random_hash);
+		    <OwnedKittiesCount<T>>::insert(&sender, new_owned_kitty_count);
+		    <OwnedKittiesIndex<T>>::insert(random_hash, owned_kitty_count);
+
+		    // Nonce update can stay here
+		    Nonce::mutate(|n| *n += 1);
+
+		    // ACTION: Move this event to the `mint()` function
+		    //Self::deposit_event(RawEvent::Created(sender, random_hash));
+
+		    Ok(())
 		}
+		
+		fn breed_kitty(origin, kitty_id_1: T::Hash, kitty_id_2: T::Hash) -> Result{
+            let sender = ensure_signed(origin)?;
+
+            ensure!(<Kitties<T>>::exists(kitty_id_1), "This cat 1 does not exist");
+            ensure!(<Kitties<T>>::exists(kitty_id_2), "This cat 2 does not exist");
+
+            let nonce = Nonce::get();
+            let random_hash = (<system::Module<T>>::random_seed(), &sender, nonce)
+                .using_encoded(<T as system::Trait>::Hashing::hash);
+
+			let payload = (<system::Module<T>>::random_seed(), &sender, 
+				<system::Module<T>>::extrinsic_index(), 
+				<system::Module<T>>::block_number());
+			let dna_hash = payload.using_encoded(blake2_128);
+
+            let kitty_1 = Self::kitty(kitty_id_1);
+            let kitty_2 = Self::kitty(kitty_id_2);
+
+            let mut final_dna = kitty_1.dna;
+            //for (i, (dna_2_element, r)) in kitty_2.dna.as_ref().iter().zip(random_hash.as_ref().iter()).enumerate() {
+            //     if r % 2 == 0 {
+            //         final_dna.as_mut()[i] = *dna_2_element;
+            //     }
+            // }			
+			for (i, (dna_2_element, r)) in kitty_2.dna.as_ref().iter().zip(dna_hash.as_ref().iter()).enumerate()  {
+				if r % 2 == 0 {
```

Comment:
the dna from kitty1 is much stronger than dna from kitty2, which makes sense, potentially we can make the strongness dynamically which means `2` can be replaced with another random number.